### PR TITLE
[IOTDB-3073][IOTDB-3095] Unescape identifier and Use antlr to parse PartialPath/Path

### PR DESCRIPTION
**Pull request by Plutooooooo**
## Description
In this PR, I use antlr to split path(String type) in the construction methods of PartialPath and Path. Before this PR, path split was done through a written split function which may lead to unexpected errors in some corner cases. After merging this PR, path is correct only if you can write it in a SQL statement, too. For example, new PartialPath("root.sg.select") will throw exception now because "select" is a key word and in SQL statement, it should be written as root.sg.\`select\`. Thus, new PartialPath("root.sg.\`select\`") is the correct form. Before this PR, in some occasion, PartialPath(String path) was called, but it didn't matter whether the path were split correctly, for we only care about the fullPath of PartialPath. In this PR, I fix problems related to this cause. Below are the cases:
1. In GroupByLevelDataSet, call PartialPath(String path, boolean needSplit) explicitly.
2. Modified  fillTargetPathTemplate in CountinousQueryTask.
3. Modified SHA256DeviceID in idtable package.

### CODE REVIEW

File: node-commons/pom.xml [hunk 1]
```diff
@@ -83,6 +83,13 @@
             <artifactId>iotdb-thrift-confignode</artifactId>
             <version>${project.version}</version>
         </dependency>
+        <!-- antlr for test -->
+        <dependency>
+            <groupId>org.apache.iotdb</groupId>
+            <artifactId>iotdb-antlr</artifactId>
+            <version>${project.version}</version>
+            <scope>compile</scope>
```
Comment (SteveYurongSu) [inline node-commons/pom.xml:91]:
```suggestion
            <scope>test</scope>
```

File: node-commons/src/main/java/org/apache/iotdb/commons/path/PartialPath.java [hunk 2]
```diff
@@ -68,8 +68,19 @@ public PartialPath(String path) throws IllegalPathException {
   }
 
   public PartialPath(String device, String measurement) throws IllegalPathException {
-    this.fullPath = device + TsFileConstant.PATH_SEPARATOR + measurement;
+    if ("".equals(device) && "".equals(measurement)) {
```
Comment (SteveYurongSu) [inline node-commons/src/main/java/org/apache/iotdb/commons/path/PartialPath.java:71]:
```suggestion
    Validator.notNull(device);
    Validator.notNull(measurement);
```

File: node-commons/src/main/java/org/apache/iotdb/commons/utils/PathUtils.java [hunk 3]
```diff
@@ -19,80 +19,51 @@
...
    * @throws IllegalPathException if path isn't correct, the exception will throw
    */
   public static String[] splitPathToDetachedPath(String path) throws IllegalPathException {
-    // NodeName is treated as identifier. When parsing identifier, unescapeJava is called.
-    // Therefore we call unescapeJava here.
-    path = StringEscapeUtils.unescapeJava(path);
-    if (path.endsWith(TsFileConstant.PATH_SEPARATOR)) {
+    if (isInnerKeyWord(path)) {
+      return new String[] {path};
+    }
```
Comment (SteveYurongSu) [inline node-commons/src/main/java/org/apache/iotdb/commons/utils/PathUtils.java:44]: isInnerKeyWord

Comment (SteveYurongSu) [reply node-commons/src/main/java/org/apache/iotdb/commons/utils/PathUtils.java:44]: Session

File: node-commons/src/main/java/org/apache/iotdb/commons/utils/PathUtils.java [hunk 4]
```diff
@@ -19,80 +19,51 @@
...
-        }
+  }
 
-        nodes.add(node);
-        // skip the '.' after '`'
-        i = endIndex + 1;
-        startIndex = endIndex + 2;
-      }
+  public static void isLegalPath(String path) throws IllegalPathException {
+    if (isInnerKeyWord(path) || "".equals(path)) {
```
Comment (SteveYurongSu) [inline node-commons/src/main/java/org/apache/iotdb/commons/utils/PathUtils.java:56]: 分离用户与内部调用

File: server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java [hunk 5]
```diff
@@ -64,7 +64,7 @@ public class IoTDBConfig {
       "org.apache.iotdb.db.conf.directories.strategy.";
   private static final String DEFAULT_MULTI_DIR_STRATEGY = "MaxDiskUsableSpaceFirstStrategy";
 
-  private static final String STORAGE_GROUP_MATCHER = "([a-zA-Z0-9_.\\-\\u2E80-\\u9FFF]+)";
+  private static final String STORAGE_GROUP_MATCHER = "([a-zA-Z0-9`_.\\-\\u2E80-\\u9FFF]+)";
```
Comment (SteveYurongSu) [inline server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java:67]: Mac Windows Linux

File: server/src/main/java/org/apache/iotdb/db/engine/cq/ContinuousQueryTask.java [hunk 6]
```diff
@@ -168,16 +168,17 @@ protected List<PartialPath> generateTargetPaths(List<Path> rawPaths) throws Ille
...
-      nodes[0] = nodes[0].substring(indexOfLeftBracket + 1);
+      fullPath = fullPath.substring(indexOfLeftBracket + 1);
     }
-    int indexOfRightBracket = nodes[nodes.length - 1].indexOf(")");
+    int indexOfRightBracket = fullPath.lastIndexOf(")");
     if (indexOfRightBracket != -1) {
-      nodes[nodes.length - 1] = nodes[nodes.length - 1].substring(0, indexOfRightBracket);
+      fullPath = fullPath.substring(0, indexOfRightBracket);
     }
+    String[] nodes = new PartialPath(fullPath).getNodes();
```
Comment (SteveYurongSu) [inline server/src/main/java/org/apache/iotdb/db/engine/cq/ContinuousQueryTask.java:181]: ?

File: server/src/main/java/org/apache/iotdb/db/metadata/path/AlignedPath.java [hunk 7]
```diff
@@ -58,26 +59,36 @@ public AlignedPath() {}
 
   public AlignedPath(String vectorPath, List<String> subSensorsList) throws IllegalPathException {
     super(vectorPath);
+    // check whether subSensor is legal
```
Comment (SteveYurongSu) [inline server/src/main/java/org/apache/iotdb/db/metadata/path/AlignedPath.java:62]: 分离用户接口 内部接口

File: tsfile/pom.xml [hunk 8]
```diff
@@ -98,6 +98,13 @@
             <version>4.0.3</version>
             <scope>test</scope>
         </dependency>
+        <!-- antlr -->
+        <dependency>
+            <groupId>org.apache.iotdb</groupId>
+            <artifactId>iotdb-antlr</artifactId>
+            <version>${project.version}</version>
+            <scope>compile</scope>
```
Comment (SteveYurongSu) [inline tsfile/pom.xml:106]:
```suggestion
```

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/common/Path.java [hunk 9]
```diff
@@ -108,14 +95,24 @@ public Path(String pathSc, boolean needSplit) {
    */
   public Path(String device, String measurement) {
     if (device == null || measurement == null) {
-      throw new IllegalArgumentException(ILLEGAL_PATH_ARGUMENT);
+      throw new PathParseException(ILLEGAL_PATH_ARGUMENT);
```
Comment (SteveYurongSu) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/read/common/Path.java:98]: 区分内部外部调用

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/common/parser/PathNodesGenerator.java [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+  public static String[] splitPathToNodes(String path) throws PathParseException {
+    try {
+      return invokeParser(path);
+    } catch (ParseCancellationException e) {
+      throw new PathParseException(path);
+    }
+  }
+
+  private static String[] invokeParser(String path) {
+    PathVisitor pathVisitor = new PathVisitor();
```
Comment (SteveYurongSu) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/read/common/parser/PathNodesGenerator.java:43]: Cache it？

File: tsfile/src/test/java/org/apache/iotdb/tsfile/read/common/PathTest.java [hunk 11]
```diff
@@ -18,32 +18,125 @@
...
+    }
+
+    try {
+      new Path("root.a*b", true);
+      fail();
+    } catch (PathParseException ignored) {
+    }
+
+    //    try {
+    //      Path path = new Path("root.*ab*", true);
```
Comment (SteveYurongSu) [inline tsfile/src/test/java/org/apache/iotdb/tsfile/read/common/PathTest.java:137]: ？

File: server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java (hunk 5 above)
Comment (Plutooooooo) [reply server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java:67]: all allow ` in file path.

File: tsfile/pom.xml (hunk 8 above)
Comment (Plutooooooo) [reply tsfile/pom.xml:106]: fixed.

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/common/parser/PathNodesGenerator.java (hunk 10 above)
Comment (Plutooooooo) [reply tsfile/src/main/java/org/apache/iotdb/tsfile/read/common/parser/PathNodesGenerator.java:43]: fixed it.

File: node-commons/src/main/java/org/apache/iotdb/commons/path/PartialPath.java [hunk 12]
```diff
@@ -79,11 +79,17 @@ public PartialPath(String[] partialNodes) {
...
    */
   public PartialPath(String path, boolean needSplit) {
     super(path, needSplit);
+    if (!needSplit) {
+      if ("".equals(path)) {
+        this.nodes = new String[] {};
+      } else {
+        this.nodes = new String[] {path};
+      }
+    }
```
Comment (SteveYurongSu) [inline node-commons/src/main/java/org/apache/iotdb/commons/path/PartialPath.java:92]: RuntimeException

File: node-commons/src/main/java/org/apache/iotdb/commons/path/PartialPath.java [hunk 13]
```diff
@@ -79,11 +79,17 @@ public PartialPath(String[] partialNodes) {
 
   /**
    * @param path path
-   * @param needSplit needSplit is basically false, whether need to be split to device and
-   *     measurement, doesn't support escape character yet.
+   * @param needSplit whether to split path to nodes.
    */
   public PartialPath(String path, boolean needSplit) {
     super(path, needSplit);
+    if (!needSplit) {
```
Comment (SteveYurongSu) [inline node-commons/src/main/java/org/apache/iotdb/commons/path/PartialPath.java:86]: Validator.requireFalse();

File: tsfile/pom.xml [hunk 14]
```diff
@@ -98,6 +98,12 @@
             <version>4.0.3</version>
             <scope>test</scope>
         </dependency>
+        <!-- antlr -->
```
Comment (SteveYurongSu) [inline tsfile/pom.xml:101]: Remove <artifactId>iotdb-antlr</artifactId> from server's dependencies.

File: tsfile/src/test/java/org/apache/iotdb/tsfile/read/common/PathTest.java (hunk 11 above)
Comment (SteveYurongSu) [reply tsfile/src/test/java/org/apache/iotdb/tsfile/read/common/PathTest.java:137]: Remove it

File: antlr/src/main/antlr4/org/apache/iotdb/db/qp/sql/IdentifierParser.g4 [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+    | TEMPLATES
+    | TIMESERIES
+    | TO
+    | TOLERANCE
+    | TOP
+    | TRACING
+    | TRIGGER
+    | TRIGGERS
+    | TTL
+    | UNLINK
```
Comment (ericpai) [inline antlr/src/main/antlr4/org/apache/iotdb/db/qp/sql/IdentifierParser.g4:161]: Are `TIME` and `TIMESTAMP` not allowed in the identifier?

Comment (Plutooooooo) [reply antlr/src/main/antlr4/org/apache/iotdb/db/qp/sql/IdentifierParser.g4:161]: Reserved words are not allowed.

File: node-commons/src/test/java/org/apache/iotdb/commons/path/PartialPathTest.java [hunk 16]
```diff
@@ -23,8 +23,496 @@
...
+    // quoted node of digits
+    PartialPath h = new PartialPath("root.sg.`111`");
+    Assert.assertEquals("root.sg.`111`", h.getFullPath());
+    nodes = new String[] {"root", "sg", "`111`"};
+    checkNodes(nodes, h.getNodes());
+
+    // quoted node of key word
+    PartialPath i = new PartialPath("root.sg.`select`");
+    Assert.assertEquals("root.sg.`select`", i.getFullPath());
+    nodes = new String[] {"root", "sg", "`select`"};
```
Comment (ericpai) [inline node-commons/src/test/java/org/apache/iotdb/commons/path/PartialPathTest.java:82]: It seems that this behavior is not as our expected.
IMO, 
```
root.sg.abc
```
and
```
root.sg.`abc`
```
should be same timeseries, which is actually stored as `String[]{"root", "sg", "abc"}`.

If a user wants to save backquote, or other special chars, he should use 

```
root.sg.`ab``.c`
```

and the actual timeseries will be
```
String[]{"root", "sg", "ab`.c"}
```

@qiaojialin @SteveYurongSu What do you think?

**Review by ericpai — CHANGES_REQUESTED** [review]

File: node-commons/src/test/java/org/apache/iotdb/commons/path/PartialPathTest.java (hunk 16 above)
Comment (qiaojialin) [reply node-commons/src/test/java/org/apache/iotdb/commons/path/PartialPathTest.java:82]: +1 for this case

Comment (Plutooooooo) [reply node-commons/src/test/java/org/apache/iotdb/commons/path/PartialPathTest.java:82]: root.sg.`abc` will be String[]{"root", "sg", "abc"}. 
root.sg.`select` will be String[]{"root", "sg", "select"}.
root.sg.`111` will be String[]{"root", "sg", "`111`"}.
root.sg.`ab``.c` will be String[]{"root", "sg", "`ab``.c`"}

File: server/src/main/java/org/apache/iotdb/db/qp/sql/IoTDBSqlVisitor.java [hunk 17]
```diff
@@ -194,6 +194,11 @@ public class IoTDBSqlVisitor extends IoTDBSqlParserBaseVisitor<Operator> {
   // for create-cq clause and select-into clause.
   private static final Pattern leveledPathNodePattern = Pattern.compile("\\$\\{\\w+}");
 
+  /** used to match node name without special characters */
+  private static final String NODE_NAME_MATCHER = "([a-zA-Z0-9_:@#${}\\u2E80-\\u9FFF]+)";
```
Comment (ericpai) [inline server/src/main/java/org/apache/iotdb/db/qp/sql/IoTDBSqlVisitor.java:198]: This is an important constant but duplicated in `ASTVisitor.java`. Could it be put into a common place?

File: server/src/main/java/org/apache/iotdb/db/qp/sql/IoTDBSqlVisitor.java [hunk 18]
```diff
@@ -2432,13 +2437,24 @@ private PartialPath convertConstantToPath(String src) throws IllegalPathExceptio
...
-    return ctx.getText();
+    return parseNodeString(ctx.getText());
   }
 
   private String parseNodeNameWithoutWildCard(IoTDBSqlParser.NodeNameWithoutWildcardContext ctx) {
-    return ctx.getText();
+    return parseNodeString(ctx.getText());
+  }
+
+  private String parseNodeString(String nodeName) {
```
Comment (ericpai) [inline server/src/main/java/org/apache/iotdb/db/qp/sql/IoTDBSqlVisitor.java:2448]: This method is duplicated.

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/common/Path.java [hunk 19]
```diff
@@ -99,19 +96,25 @@ public Path(String device, String measurement) {
...
     // use PathNodesGenerator to check whether path is legal.
     if (!"".equals(device) && !"".equals(measurement)) {
-      String fullPath = device + TsFileConstant.PATH_SEPARATOR + measurement;
-      PathNodesGenerator.checkPath(fullPath);
+      String path = device + TsFileConstant.PATH_SEPARATOR + measurement;
+      String[] nodes = PathNodesGenerator.splitPathToNodes(path);
+      this.device = transformNodesToString(nodes, nodes.length - 1);
+      this.measurement = nodes[nodes.length - 1];
+      this.fullPath = transformNodesToString(nodes, nodes.length);
     } else if (!"".equals(device)) {
```
Comment (ericpai) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/read/common/Path.java:104]: Maybe we can use `StringUtils.isEmpty()` instead here

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/common/parser/PathVisitor.java [hunk 20]
```diff
@@ -21,11 +21,18 @@
...
+import org.apache.iotdb.tsfile.common.constant.TsFileConstant;
+
+import org.apache.commons.lang3.StringUtils;
 
 import java.util.List;
+import java.util.regex.Pattern;
 
 public class PathVisitor extends PathParserBaseVisitor<String[]> {
 
+  private static final String NODE_NAME_MATCHER = "([a-zA-Z0-9_:@#${}\\u2E80-\\u9FFF]+)";
```
Comment (ericpai) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/read/common/parser/PathVisitor.java:33]: Duplicated constant here.

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/parser/ASTVisitor.java [hunk 21]
```diff
@@ -1170,10 +1187,8 @@ private String parseStringLiteralInInsertValue(String src) {
   }
 
   private String parseIdentifier(String src) {
-    if (2 <= src.length() && src.charAt(0) == '`' && src.charAt(src.length() - 1) == '`') {
-      String unescaped = StringEscapeUtils.unescapeJava(src.substring(1, src.length() - 1));
-      // replace `` with `
-      return unescaped.replace("``", "`");
+    if (src.startsWith("`") && src.endsWith("`")) {
```
Comment (ericpai) [inline server/src/main/java/org/apache/iotdb/db/mpp/plan/parser/ASTVisitor.java:1190]: Use a common constant instead of "`" and  "``" directly.

**Review by ericpai — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/apache/iotdb/db/qp/sql/IoTDBSqlVisitor.java (hunk 18 above)
Comment (Plutooooooo) [reply server/src/main/java/org/apache/iotdb/db/qp/sql/IoTDBSqlVisitor.java:2448]: NodeName and NodeNameWithoutWildCard could be different in the future, maybe it is better this way?

File: server/src/main/java/org/apache/iotdb/db/qp/sql/IoTDBSqlVisitor.java (hunk 17 above)
Comment (Plutooooooo) [reply server/src/main/java/org/apache/iotdb/db/qp/sql/IoTDBSqlVisitor.java:198]: fixed. Removed to TsFileConstant.

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/common/Path.java (hunk 19 above)
Comment (Plutooooooo) [reply tsfile/src/main/java/org/apache/iotdb/tsfile/read/common/Path.java:104]: fixed.

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/parser/ASTVisitor.java (hunk 21 above)
Comment (Plutooooooo) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/parser/ASTVisitor.java:1190]: fixed.

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/common/parser/PathVisitor.java (hunk 20 above)
Comment (Plutooooooo) [reply tsfile/src/main/java/org/apache/iotdb/tsfile/read/common/parser/PathVisitor.java:33]: fixed.

File: server/src/main/java/org/apache/iotdb/db/qp/sql/IoTDBSqlVisitor.java (hunk 18 above)
Comment (ericpai) [reply server/src/main/java/org/apache/iotdb/db/qp/sql/IoTDBSqlVisitor.java:2444]: Got it, let's leave them separated.

**Review by ericpai — APPROVED** [review]

**Comment by ericpai** [comment]
LGTM. Excellent work!

**Pull request merged**